fpdf/drawing.py

```diff
@@ -0,0 +1,1023 @@
+
+    @contextmanager
+    def transform_group(self, transform):
+        """
+        Apply the provided transforms to all points added within this context.
+        """
+        with self._new_graphics_context() as ctxt:
+            ctxt.transform = transform
+            yield self
+
+    def _push_drawing_sequence(self, item):
+        if self._starter_move is not None:
+            self._closed = False
+
+            self._graphics_context.add_item(self._starter_move)
+            self._close_context = self._graphics_context
+            self._starter_move = None
+
+        self._graphics_context.add_item(item)
+
+    def rectangle(self, x, y, w, h):
+        """
+        Append a rectangle as a closed subpath to the current path.
+
+        The rectangle has all of the same constraints that regular subpaths do.
+        """
+        self._close()
+        self._push_drawing_sequence(Rectangle(Point(x, y), Point(w, h)))
+
+        return self
+
+    def move_to(self, x, y):
+        self._close()
+        self._starter_move = Move(Point(x, y))
+
+    def move_relative(self, x, y):
+        self._close()
+        self._starter_move = RelativeMove(Point(x, y))
+
+    def line_to(self, x, y):
+        point = Point(x, y)
+        self._push_drawing_sequence(Line(point))
+        return self
+
+    def line_relative(self, dx, dy):
+        offset = Point(dx, dy)
+        self._push_drawing_sequence(RelativeLine(offset))
+        return self
+
+    def curve_to(self, x1, y1, x2, y2, x3, y3):
+        ctrl1 = Point(x1, y1)
+        ctrl2 = Point(x2, y2)
+        end = Point(x3, y3)
+
+        self._push_drawing_sequence(BezierCurve(ctrl1, ctrl2, end))
+        return self
+
+    def curve_relative(self, dx1, dy1, dx2, dy2, dx3, dy3):
+        """
+        Append a cubic Bézier curve whose points are each expressed relative to the
+        preceding point.
+
+        E.g. with a start point of (0, 0), given (1, 1), (2, 2), (3, 3), the output
+        curve would have the points:
+
+        (0, 0) c1 (1, 1) c2 (3, 3) e (6, 6)
+        """
+        c1d = Point(dx1, dy1)
+        c2d = Point(dx2, dy2)
+        end = Point(dx3, dy3)
+
+        self._push_drawing_sequence(RelativeBezierCurve(c1d, c2d, end))
+        return self
+
+    def quadratic_bezier_to(self, x1, y1, x2, y2):
+        """
+        Append a cubic Bézier curve mimicking the specified quadratic Bézier curve.
+        """
+        ctrl = Point(x1, y1)
+        end = Point(x2, y2)
+        self._push_drawing_sequence(QuadraticBezierCurve(ctrl, end))
+        return self
+
+    def quadratic_bezier_relative(self, x1, y1, x2, y2):
+        """
+        Append a cubic Bézier curve mimicking the specified quadratic Bézier curve.
+        """
+        ctrl = Point(x1, y1)
+        end = Point(x2, y2)
+        self._push_drawing_sequence(RelativeQuadraticBezierCurve(ctrl, end))
+        return self
+
+    def close(self):
+        self.closed = True
```

Comment:
What if `self.closed` is already `True`?
Also, shouldn't this method call `self._close()`?

Maybe method naming could be slightly improved here in order to make the roles of `_close()` & `close()` more obvious

Reply:
`self._close()` should probably be named something more like `self._auto_close_path_if_appropriate()`. To elaborate on this: PDF offers some paint styles that automatically close the path (i.e. they insert an implicit `h` just before painting). Path closing affects only the stroke, as fill is always drawn as if the path had been closed. However, because a single painted path can be composed of multiple closed shapes (e.g. `0 0 m 6 0 l 6 6 l 0 6 l 1 1 m 0 5 l 5 5 l 5 0 l s` is a square inside a square) the `stroke_closed` painting style will only actually close the last path before painting, which in the example given means the outer square (specified first) will actually be rendered as a C shape rather than a closed box. To exercise the principle of least astonishment, we try to automatically insert close-path operators when creating a new shape within a single painting operation that is using an automatic closing paint style (so the above example would become `0 0 m 6 0 l 6 6 l 0 6 l h 1 1 m 0 5 l 5 5 l 5 0 l s`). This method does the bookkeeping to avoid emitting unnecessary path close operators.

I should probably make it insert an `ImplicitClose` operator that can emit `h` or nothing on rendering so that a path can be reused with different paint styles and do the right thing at render time rather than at composition time.

I added the `self.close()` API later when working on the SVG converter, as SVG also has an explicit path close operator (`Z`), which makes sense when generating paths from an existing source (it also makes sense to allow users to explicitly close paths if they want to, anyway).

Reply:
this is now named `_insert_implicit_close_if_open` which I hope is sufficiently self-explanatory. It will still benefit from a docstring, however.